### PR TITLE
Develop qmc finance

### PR DESCRIPTION
**Pull request by ggraoigr**
Pull request to merge with develop branch

### CODE REVIEW

File: qmcpy/integrand/LookBackOption.py
Comment (alegresor) [inline qmcpy/integrand/LookBackOption.py:47]: Do not override this method. 
The transform to a geometric brownian motion is already set to `self.s` in the `g` method below

File: qmcpy/stopping_criterion/.ipynb_checkpoints/_cub_qmc_ld_g-checkpoint.py
Comment (alegresor) [inline qmcpy/stopping_criterion/.ipynb_checkpoints/_cub_qmc_ld_g-checkpoint.py:1]: How is this different than the existing `_CubQMCLDG` class?
Please remove if it is a duplicate or merge changes into the existing class.

File: qmcpy/stopping_criterion/Fixed_multilevel.py
Comment (alegresor) [inline qmcpy/stopping_criterion/Fixed_multilevel.py:60]: I'm not sure how this stopping criterion works. The `MLQMCData` object is for adaptive multi-level problems and the adaptive mechanisms are used below e.g. in `self.data._add_level()`. However, this seems to be written for fixed multi-level problems? 

Do you have a writeup or reference on how this works?

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (alegresor) [inline qmcpy/stopping_criterion/_cub_qmc_ld_g.py:8]: move this to within the method performing the plotting. This makes it so matplotlib is only imported if the user wants to plot something, but does not make it a requirement of QMCPy.

**Review by alegresor — CHANGES_REQUESTED** [review]
Thanks for putting this together all! 

Left a number of comments below. 

I think the american option needs some extra thought as to how we should update previous samples in a stopping criterion. 

Please provide a reference for the fixed multi-level stopping criterion. I think the current algorithm is actually adaptive. 

The lookback option seems fine. 

The density estimation method should be moved to the accumulate data class, be refactored to make plotting optional, and be allowed to take in samples at which we evaluate the combined discrete/continuous KDE for the user. 

I'll take a closer look at these files, including the notebooks, when we are closer to merging. 

If you'd like to make a separate PR with just the density estimation and lookback options parts I think we can get that merged quicker than the american option and fixed multi-level parts

File: qmcpy/integrand/AmericanOption.py
Comment (rvare) [inline qmcpy/integrand/AmericanOption.py:64]: I want to note that on line 64 which is `exercise=in_money_values` could have potential problems later on. This statement does not make a copy the list but instead has `exercise` point to the same list that `in_money_values` also points to. At the moment it seems that the code works as it should due to pure luck, but this could break at a later time. However, if this was intended, then disregard this comment.

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (rvare) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py:169]: I will delete these statements.

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (rvare) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py:180]: What I'll do is make an if statement and change the method header to have a Boolean flag to make plotting optional.

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (rvare) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py:193]: I'll make it so it returns `fig` and `ax`. However, there is an `ax1` and `ax2`, do both of these need to be returned or only one?

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (rvare) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py:8]: Since I plan on making an if-statement that handles plotting, I'll move that there as I make those changes.

File: qmcpy/integrand/AmericanOption.py
Comment (rvare) [reply qmcpy/integrand/AmericanOption.py:64]: For some reason GitHub highlighted the code above the line I've mentioned as well, even though I highlighted the code that was underneath it because it's also related. Not sure why GitHub messed that up.

File: qmcpy/integrand/LookBackOption.py
Comment (rvare) [reply qmcpy/integrand/LookBackOption.py:47]: I'll try to delete this once I finish refactoring the all the options to use the abstract option class. I don't want to make too many changes and have everything start breaking because of it. Once this branch gets merged with mine, then I'll remove it.

**Comment by rvare** [comment]
Changed the base branch from `main` to `develop`.

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (alegresor) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py]: Try to use `fig,ax = pyplot.subplots(ncols=2)` so that `ax` is a `numpy.ndarray` of axes which may be returned

File: qmcpy/integrand/AmericanOption.py
Comment (alegresor) [reply qmcpy/integrand/AmericanOption.py:64]: Maybe you don't need the `exercise` variable and can just use `in_money_values` instead. Otherwise use `numpy.copy`

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (rvare) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py]: I've been dealing with an issue with this possible solution. When doing this, two subplots are made and not one. It seems when you do `ncols=2` two separate plots are made, but they need to be overlaid with each other. I'm still looking into this, but I figure this was something important to note.

File: qmcpy/stopping_criterion/__init__.py
Comment (rvare) [reply qmcpy/stopping_criterion/__init__.py:13]: I will remove the import statement.

File: qmcpy/integrand/AmericanOption.py
Comment (rvare) [reply qmcpy/integrand/AmericanOption.py:64]: Just to reduce the risk of something breaking, I went ahead and used `numpy.copy` so that it would mostly stay the same.

File: qmcpy/stopping_criterion/.ipynb_checkpoints/_cub_qmc_ld_g-checkpoint.py
Comment (aaditj1962161) [reply qmcpy/stopping_criterion/.ipynb_checkpoints/_cub_qmc_ld_g-checkpoint.py:1]: Looking at it, I myself don't see any changes except the ones Richard recently made. We can probably delete this.

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (aaditj1962161) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py]: Can you clarify this a bit more? Me and Richard are confused as to what you exactly mean.

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (aaditj1962161) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py]: How about we do `axes = array([ax1, ax2]) ` and then return axes? If we use the subplots than we'll get two seperate plots and I think twinx() is the best way to have one plot share the x axes and have two seperate y axes?

Comment (rvare) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py]: That makes sense. I originally thought I had to use the object returned from `subplots`, but this makes more sense. I'll see if I can make this change today.

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (rvare) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py:163]: I'm not sure what class that is nor how that would work. Can you give some clarification about what you had in mind?

File: qmcpy/stopping_criterion/.ipynb_checkpoints/_cub_qmc_ld_g-checkpoint.py
Comment (rvare) [reply qmcpy/stopping_criterion/.ipynb_checkpoints/_cub_qmc_ld_g-checkpoint.py:1]: I'll go ahead and delete this. However, I'll probably delete the whole folder otherwise will just be tracking an empty folder at that point.

File: qmcpy/stopping_criterion/Fixed_multilevel.py
Comment (aaditj1962161) [reply qmcpy/stopping_criterion/Fixed_multilevel.py:60]: Guillem doesn't remember about this. So should we just ignore this and not merge it?

File: qmcpy/integrand/LookBackOption.py
Comment (aaditj1962161) [inline qmcpy/integrand/LookBackOption.py:1]: should it not be `return LookBackOption()` instead of `return AmericanOption()`in the spawn function

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (alegresor) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py:163]: Start by putting it in `qmcpy/accumulate_data/ld_transform_data.py`, but later we would like it to be general enough to support all stopping criterion

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (alegresor) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py]: Instead of returning the `kde` object for the continuous part, let the user pass in samples at which to evaluate the full KDE. Be careful as there is both the continuous part and discrete parts of the KDE.

File: qmcpy/stopping_criterion/Fixed_multilevel.py
Comment (alegresor) [reply qmcpy/stopping_criterion/Fixed_multilevel.py:60]: Sure, that works

File: qmcpy/integrand/LookBackOption.py
Comment (alegresor) [reply qmcpy/integrand/LookBackOption.py:1]: Yes, spawn should return a new instance of the class its spawning with the same parameters

File: qmcpy/stopping_criterion/Fixed_multilevel.py
Comment (aaditj1962161) [reply qmcpy/stopping_criterion/Fixed_multilevel.py:60]: I added the following to `.gitignore`: `stopping_criterion/Fixed_multilevel.py`. Did it work?

File: qmcpy/integrand/LookBackOption.py
Comment (rvare) [reply qmcpy/integrand/LookBackOption.py:1]: I can make this change today and push it once it's done.

File: qmcpy/integrand/LookBackOption.py
Comment (rvare) [reply qmcpy/integrand/LookBackOption.py:47]: I went a head and fixed this. `LookBackOption` and `AmericanOption` no longer override `f` anymore.

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (aaditj1962161) [reply qmcpy/stopping_criterion/_cub_qmc_ld_g.py]: Updated. Pls check

File: qmcpy/stopping_criterion/Fixed_multilevel.py
Comment (aaditj1962161) [reply qmcpy/stopping_criterion/Fixed_multilevel.py]: Deleted the file

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (alegresor) [inline qmcpy/stopping_criterion/_cub_qmc_ld_g.py:199]: Return these values and let them user print them if they choose

File: qmcpy/stopping_criterion/_cub_qmc_ld_g.py
Comment (alegresor) [inline qmcpy/stopping_criterion/_cub_qmc_ld_g.py:190]: `densities.ndim==1` yet we are assigning an element to `kde_discrete` which is a vector? I think you need to find the index `j` where `values[i]` is in in `repeated_values` then set `densities[i] = kde_discrete[j]`. Let me know if I'm misunderstanding

**Review by alegresor — CHANGES_REQUESTED** [review]

File: demos/talk_paper_demos/SURE_2024_Demos/LookBack_Adaptive_Tests.ipynb
Comment (aaditj1962161) [inline demos/talk_paper_demos/SURE_2024_Demos/LookBack_Adaptive_Tests.ipynb:1]: Would this need to be updated a bit considering the density function has been updated

Comment (rvare) [reply demos/talk_paper_demos/SURE_2024_Demos/LookBack_Adaptive_Tests.ipynb:1]: Most likely. I'll take care of updating it which shouldn't take me too long.

**Pull request closed without merging**
